apps/studio/src/server/modules/user/user.service.ts

```diff
@@ -11,3 +20,175 @@ export const isUserDeleted = async (email: string) => {
+  if (!isEmail(email)) {
+    throw new TRPCError({
+      code: "BAD_REQUEST",
+      message: "Invalid email",
+    })
+  }
+
+  validateEmailRoleCombination({ email, role })
+
+  const isWhitelisted = await isEmailWhitelisted(email)
+  if (!isWhitelisted) {
+    throw new TRPCError({
+      code: "FORBIDDEN",
+      message: "There are non-gov.sg domains that need to be whitelisted.",
+    })
+  }
+
+  const executeInTransaction = async (tx: SafeKysely) => {
+    const user = await tx
+      .insertInto("User")
+      .values({
+        id: cuid2.createId(),
+        email,
+        name,
+        phone,
+      })
+      .onConflict((oc) =>
+        oc
+          .column("email")
+          .doUpdateSet((eb) => ({ email: eb.ref("excluded.email") })),
+      )
+      .returning(["id", "email", "name", "phone", "deletedAt"])
+      .executeTakeFirstOrThrow()
+
+    // Temporary test to ensure that we don't allow creating a user that was deleted before
+    // We prevent this as there's no complete audit trail of what happened to the user
+    // TODO: Remove this after audit logging is implemented
+    if (user.deletedAt) {
+      throw new TRPCError({
+        code: "CONFLICT",
+        message: "User was deleted before. Contact support to restore.",
+      })
+    }
+
+    // unique constraint (@@unique([userId, siteId, resourceId, role]))
+    // does not work if one column is NULL e.g. resourceId
+    // Thus we have to check on the application layer
+    const existingResourcePermission = await tx
+      .selectFrom("ResourcePermission")
+      .where("userId", "=", user.id)
+      .where("siteId", "=", siteId)
+      .where("resourceId", "is", null) // because we are granting site-wide permissions
+      .selectAll()
+      .executeTakeFirst()
+
+    if (existingResourcePermission) {
+      throw new TRPCError({
+        code: "CONFLICT",
+        message: "User and permissions already exists",
+      })
+    }
+
+    const resourcePermission = await tx
+      .insertInto("ResourcePermission")
+      .values({
+        userId: user.id,
+        siteId,
+        role,
+      })
+      .returning(["userId", "siteId", "role"])
+      .executeTakeFirstOrThrow()
+
+    return {
+      user,
+      resourcePermission,
+    }
+  }
+
+  if (trx) {
+    return await executeInTransaction(trx)
+  }
+
+  return await db.transaction().execute(executeInTransaction)
+}
+
+interface GetUsersQueryProps {
+  siteId: number
+  getIsomerAdmins: boolean
+}
+
+interface RankedResourcePermission extends ResourcePermission {
+  rn: number
+}
+
+export const getUsersQuery = ({
+  siteId,
+  getIsomerAdmins,
+}: GetUsersQueryProps) => {
+  return db
+    .with("ActiveResourcePermission", (qb) =>
+      qb
+        .selectFrom(
+          sql<RankedResourcePermission>`(
+          SELECT *,
+            ROW_NUMBER() OVER (
+              PARTITION BY "userId", "siteId", "resourceId"
+              ORDER BY CASE 
+                WHEN role = ${RoleType.Admin} THEN 1
```

Comment:
why do we need the role ranking here? each user should have exactly 1 role per site right?

Reply:
same issue as non-uniqueness on null value https://github.com/opengovsg/isomer/pull/1118#discussion_r1978864412

Reply:
Ah right... Hmm, that's an issue as it complicates our queries a lot

Might be out of scope for this PR, but can we do a unique constraint with COALESCE?

Reply:
We should have used something like -1 for global resource perms

Reply:
I mean smth like 

```
CREATE UNIQUE INDEX unique_user_site_role
ON "ResourcePermission" (userId, siteId, COALESCE(resourceId, -1), role);

```

This will automatically make all resourceId which is NULL to -1

Reply:
that might be possible as a quick hack, will look into it.

though im quite liking @njenwei [idea](https://github.com/opengovsg/isomer/pulls?q=assignee%3Aadriangohjw+is%3Aopen+%22USER+MGMT%22+assignee%3Aadriangohjw+is%3Aopen+%22USER+MGMT%22++draft%3Afalse+-review%3Aapproved+) of having a separate table `SitePermission` - it feels more self-explanatory and straightforward. Might be something to consider

Reply:
Yea it might be cleaner, but might be too much to handle in this PR? Also we need to backfill for existing sites we have during deployment. Check out the effort needed - are there places in our codebase to update checks etc

Reply:
yup it's too much to handle in this PR, esp. since this is kinda being used everywhere and we don't have much test coverage on other parts of the app haha.

will add this to a backlog

Reply:
Sounds good, if so maybe the unique constraint temporarily can simplify your code till we decouple site vs resource perms

Reply:
fixed in https://github.com/opengovsg/isomer/pull/1118/commits/c5dc7baa0cae9ddddb21d944ee10c2ce2d33b524